controller/work_item_link.go

```diff
@@ -100,107 +107,67 @@ func getTypesOfLinks(ctx *workItemLinkContext, linksDataArr []*app.WorkItemLinkD
 
 // getWorkItemsOfLinks returns an array of distinct work items as they appear as
 // source or target in the given work item links.
-func getWorkItemsOfLinks(ctx *workItemLinkContext, linksDataArr []*app.WorkItemLinkData) ([]*app.WorkItem, error) {
-	// Build our "set" of distinct work item IDs already converted as strings
-	workItemIDMap := map[uuid.UUID]bool{}
+func getWorkItemsOfLinks(ctx context.Context, appl application.Application, req *http.Request, linksDataArr []*app.WorkItemLinkData) ([]*app.WorkItem, error) {
+	// Build our "set" of distinct work item IDs
+	idMap := map[uuid.UUID]struct{}{}
+	idArr := []uuid.UUID{}
 	for _, linkData := range linksDataArr {
-		workItemIDMap[linkData.Relationships.Source.Data.ID] = true
-		workItemIDMap[linkData.Relationships.Target.Data.ID] = true
-	}
-	// Now include the optional work item data in the work item link "included" array
-	workItemArr := []*app.WorkItem{}
-	for workItemID := range workItemIDMap {
-		wi, err := ctx.Application.WorkItems().LoadByID(ctx.Context, workItemID)
-		if err != nil {
-			return nil, errs.WithStack(err)
+		src := linkData.Relationships.Source.Data.ID
+		tgt := linkData.Relationships.Target.Data.ID
+		if _, ok := idMap[src]; !ok {
+			idMap[src] = struct{}{}
+			idArr = append(idArr, src)
+		}
+		if _, ok := idMap[tgt]; !ok {
+			idMap[tgt] = struct{}{}
+			idArr = append(idArr, tgt)
 		}
-		workItemArr = append(workItemArr, ConvertWorkItem(ctx.Request, *wi))
 	}
-	return workItemArr, nil
-}
 
-// getCategoriesOfLinkTypes returns an array of distinct work item link
-// categories for the given work item link types
-func getCategoriesOfLinkTypes(ctx *workItemLinkContext, linkTypeDataArr []*app.WorkItemLinkTypeData) ([]*app.WorkItemLinkCategoryData, error) {
-	// Build our "set" of distinct category IDs already converted as strings
-	catIDMap := map[uuid.UUID]bool{}
-	for _, linkTypeData := range linkTypeDataArr {
-		catIDMap[linkTypeData.Relationships.LinkCategory.Data.ID] = true
-	}
-	// Now include the optional link category data in the work item link "included" array
-	catDataArr := []*app.WorkItemLinkCategoryData{}
-	for catID := range catIDMap {
-		modelCategory, err := ctx.Application.WorkItemLinkCategories().Load(ctx.Context, catID)
+	// Fetch all work items specified in the array
+	res := []*app.WorkItem{}
+	for _, id := range idArr {
+		wi, err := appl.WorkItems().LoadByID(ctx, id)
```

Comment:
isn't there a way to avoid running `len(idArr)` (`2` ?) times this `LoadByID` method ? Can't we have something more efficient at the repository level instead ?

Reply:
This is just old code and I didn't want to put such optimization inside this PR. I mean, I started writing something up, that I'm happy to make a PR of but please not in this one. My point is: your critique is correct but its not new code.

Reply:
ok, so let's open another issue, and make sure this is addressed, ok ?